server/app/filters/ApiKeyUsageFilter.java

```diff
@@ -0,0 +1,74 @@
+package filters;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import auth.ProfileUtils;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.Executor;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import play.api.libs.concurrent.AkkaSchedulerProvider;
+import play.mvc.EssentialAction;
+import play.mvc.EssentialFilter;
+import play.mvc.Http;
+import scala.concurrent.ExecutionContext;
+import services.apikey.ApiKeyService;
+
+/**
+ * This filter looks for requests with paths that begin with /api and record the usage data for the
+ * relevant API key if it has one. The API key update is scheduled to run in a background through to
+ * reduce latency and to ensure issues with recording the usage do not cause API requests to fail.
+ */
+public class ApiKeyUsageFilter extends EssentialFilter {
+
+  private final AkkaSchedulerProvider akkaSchedulerProvider;
+  private final Provider<ApiKeyService> apiKeyServiceProvider;
+  private final Executor exec;
+  private final Provider<ProfileUtils> profileUtilsProvider;
+
+  @Inject
+  public ApiKeyUsageFilter(
+      AkkaSchedulerProvider akkaSchedulerProvider,
+      Provider<ApiKeyService> apiKeyServiceProvider,
+      Executor exec,
+      Provider<ProfileUtils> profileUtilsProvider) {
+    this.akkaSchedulerProvider = checkNotNull(akkaSchedulerProvider);
+    this.apiKeyServiceProvider = checkNotNull(apiKeyServiceProvider);
+    this.exec = checkNotNull(exec);
+    this.profileUtilsProvider = checkNotNull(profileUtilsProvider);
+  }
+
+  @Override
+  public EssentialAction apply(EssentialAction next) {
+    return EssentialAction.of(
+        (Http.RequestHeader request) ->
+            next.apply(request)
+                .map(
```

Comment:
Not blocking: Does an exception thrown in a filter presumably block a subsequent filter from running?

Given that our semantics for this are best-effort, might it make sense to wrap this whole stanza in a try/catch (and log) so that failure to retrieve the API key / schedule the background work doesn't block the response from being sent to the client?

Reply:
done